src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -2228,6 +2232,62 @@ impl BookTest {
         // Books often have feature-gated example text.
         rustbook_cmd.env("RUSTC_BOOTSTRAP", "1");
         rustbook_cmd.env("PATH", new_path).arg("test").arg(path);
+
+        // Books may also need to build dependencies. For example, `TheBook` has
+        // code samples which use the `trpl` crate. For the `rustdoc` invocation
+        // to find them them successfully, they need to be built first and their
+        // paths used to generate the
+        let libs = if !self.dependencies.is_empty() {
+            let mut lib_paths = vec![];
+            for dep in self.dependencies {
+                let mode = Mode::ToolRustc;
```

Comment:
This seems wrong. I don't see how `src/doc/book/packages/trpl` depends on rustc crates.

Reply:
Thanks—should it be `Mode::ToolBootstrap` or something else?

Reply:
Yeah, should be `Mode::ToolBootstrap`.

Reply:
When I set it to `Mode::ToolBootstrap`, testing the book fails like this:

```sh
$ ./x test src/doc/book
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
thread 'main' panicked at src/core/builder/cargo.rs:518:9:
assertion failed: !use_snapshot || stage == 0 || self.local_rebuild
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:00:00
```

Using `Mode::ToolRustc` works correctly, which is why I ended up here. I’m not sufficiently familiar (and my reading has not yet been sufficiently illuminating!) with the code here to figure out why `ToolBootstrap` doesn’t work.

Reply:
It's expecting stage 0 compiler for `ToolBootstrap` as explained here:

https://github.com/rust-lang/rust/blob/e473783d90e2289b8a97575fa60d6315f0a318eb/src/bootstrap/src/lib.rs#L226-L232

and the following patch does that

```diff
diff --git a/src/bootstrap/src/core/build_steps/test.rs b/src/bootstrap/src/core/build_steps/test.rs
index 24e68d2c923..9cebd46e4d0 100644
--- a/src/bootstrap/src/core/build_steps/test.rs
+++ b/src/bootstrap/src/core/build_steps/test.rs
@@ -2242,10 +2242,10 @@ fn run_ext_doc(self, builder: &Builder<'_>) {
             for dep in self.dependencies {
                 let mode = Mode::ToolBootstrap;
                 let target = builder.config.build;
-                // CHECKME: is this correct, or should it be using `builder::Cargo::new`?
+                let compiler = builder.compiler(0, target);
                 let cargo = tool::prepare_tool_cargo(
                     builder,
-                    self.compiler,
+                    compiler,
                     mode,
                     builder.config.build,
                     Kind::Build,
```

Reply:
Incorporated that by setting it when saving the compiler instance (see e5194ad2).